### PR TITLE
✨ Add MATCH macro

### PR DESCRIPTION
**Pull request by jonathantyng-amp**
Add a $MATCH macro. https://github.com/ampproject/amphtml/issues/23017

`$MATCH(str, regex, opt_index)`

Applies the [match function
](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match) to the given `str` with the given `regex`.

Returns the matching group declared by `opt_index`, where `opt_index=0` indicates the full match. Defaults to 0. If there is no match for the given index, empty string is returned

E.g.
`$MATCH(test.123, test\.(\d+), 0)`     ->     `test.123`
`$MATCH(test.123, test\.(\d+), 1)`     ->     `123`
`$MATCH(test.test, test\.(\d+), 1)`     ->     ``

cc @zhouyx @zikas

### CODE REVIEW

File: extensions/amp-analytics/0.1/variables.js [hunk 1]
```diff
@@ -141,6 +141,34 @@ function replaceMacro(string, matchPattern, opt_newSubStr) {
   return string.replace(regex, opt_newSubStr);
 }
 
+/**
+ * Applies the match function to the given string with the given regex
+ * @param {string} string input to be replaced
+ * @param {string} matchPattern string representation of regex pattern
+ * @param {string=} opt_indexStr the matching group to return. Index of 0
```
Comment (torch2424) [inline extensions/amp-analytics/0.1/variables.js]: nit: can we rename this to `opt_matchingGroupIndexStr`? 😄

File: extensions/amp-analytics/0.1/variables.js [hunk 2]
```diff
@@ -141,6 +141,34 @@ function replaceMacro(string, matchPattern, opt_newSubStr) {
...
+    user().warn(TAG, 'MATCH macro must have two or more arguments');
+  }
+
+  let index = 0;
+  if (opt_indexStr) {
+    index = parseInt(opt_indexStr, 10);
+  }
+
+  // if given a non-number or negative number
+  if (!index || index < 0) {
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/variables.js]: nit: It'd be great if we can throw a user error if index is invalid. (NaN or negative)

**Comment by jonathantyng-amp** [comment]
cc @zhouyx ready for merge!

**Comment by zhouyx** [comment]
Merged! Thank you for the supporting the feature!

**Pull request merged**